### ISSUE TITLE
VS2015 hangs after closing NTVS project

### ISSUE THREAD

**Issue by eladnava** [Microsoft/nodejstools#467]
Hi,
First off, I want to say that NTVS is simply amazing! It's the most advanced way to develop and debug Node.js apps as of today. The IntelliSense is so nice to have! And so is the "edit-and-continue". No other IDE is able to match these features, and I tested a lot of them. IntelliSense for JS is extremely hard to pull off. Well done.

One thing that happens randomly is VS2015 freezing when closing a NTVS solution. It happens about 80% of the time. Everything works great until I'm done with the project and want to close it. VS will hang and won't exit even after minutes. Only a process kill via Task Manager to `devenv.exe` can close VS.

Is this a known issue? I tried searching and couldn't find it.

Thanks for your support,
Elad

**Comment by mousetraps** [comment]
Glad you like it!

Perhaps there are a lot of analysis objects in memory that are being serialized and saved to disk. Would you mind providing your package.json file so we can try to reproduce the issue?

**Comment by eladnava** [comment]
Absolutely.

	{
	  "name": "App",
	  "main": "server.js",
	  "version": "1.0.0",
	  "dependencies": {
		"async": "^0.9.0",
		"body-parser": "^1.10.1",
		"cookie-parser": "^1.3.4",
		"crypto": "0.0.3",
		"express": "~4.9.0",
		"mongoose": "^4.1.7",
		"mqtt": "^0.3.13",
		"nodemailer": "^1.3.4",
		"nodemailer-ses-transport": "^1.2.0",
		"path": "^0.11.14",
		"request": "^2.51.0",
		"stripe": "^3.3.4",
		"validator": "^4.0.5"
	  },
	  "devDependencies": {
		"jshint": "^2.8.0"
	  }
	}

**Comment by mousetraps** [comment]
After installing all of the packages you described, playing around with it a bit (waiting some time was required - I was not able to reproduce immediately), running into issue #505... I managed to reproduce the issue. 

```
Microsoft.NodejsTools.dll!Microsoft.NodejsTools.Intellisense.VsProjectAnalyzer.AnalysisQueue.Worker(object threadStarted) Line 215	C#
```

```csharp
WaitHandle.SignalAndWait(
    _analyzer._queueActivityEvent,
    _workEvent
);
```

It appears that the wait sometimes fails to complete during project close.

**Comment by eladnava** [comment]
Well done! Let me know if you need anything else, or if I can somehow help fix it.

**Comment by mousetraps** [comment]
And update / braindump on this... when closing the project, a cancellation is requested using a cancellation token.

After AnalysisUnit.Analyze, however, the cancellation token is ignored, and NTVS first completes analysis on the current analysis unit. 
https://github.com/Microsoft/nodejstools/blob/master/Nodejs/Product/Analysis/Analysis/Analyzer/AnalysisUnit.cs#L174-L176

We could pass the cancellation unit down further and cancel analysis during project close, but that would only hide the problem (in particular, analysis would just stall while the project is active in Visual Studio.)

In this case, the callstack looks something like:
```
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.GlobalBuilder.DefineProperties(Microsoft.NodejsTools.Analysis.Values.FunctionValue func, Microsoft.NodejsTools.Parsing.Node node, Microsoft.NodejsTools.Analysis.AnalysisUnit unit, Microsoft.NodejsTools.Analysis.IAnalysisSet this, Microsoft.NodejsTools.Analysis.IAnalysisSet[] args) Line 1325	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Values.SpecializedFunctionValue.Call(Microsoft.NodejsTools.Parsing.Node node, Microsoft.NodejsTools.Analysis.AnalysisUnit unit, Microsoft.NodejsTools.Analysis.IAnalysisSet this, Microsoft.NodejsTools.Analysis.IAnalysisSet[] args) Line 177	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.AnalysisSetExtensions.Call(Microsoft.NodejsTools.Analysis.IAnalysisSet self, Microsoft.NodejsTools.Parsing.Node node, Microsoft.NodejsTools.Analysis.AnalysisUnit unit, Microsoft.NodejsTools.Analysis.IAnalysisSet this, Microsoft.NodejsTools.Analysis.IAnalysisSet[] args) Line 75	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.ExpressionEvaluator.EvaluateCall(Microsoft.NodejsTools.Analysis.Analyzer.ExpressionEvaluator ee, Microsoft.NodejsTools.Parsing.Node node) Line 294	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.ExpressionEvaluator.EvaluateWorker(Microsoft.NodejsTools.Parsing.Node node) Line 118	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.ExpressionEvaluator.Evaluate(Microsoft.NodejsTools.Parsing.Expression node) Line 51	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.DDG.Walk(Microsoft.NodejsTools.Parsing.ExpressionStatement node) Line 143	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Parsing.ExpressionStatement.Walk(Microsoft.NodejsTools.Parsing.AstVisitor visitor) Line 46	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.DDG.Walk(Microsoft.NodejsTools.Parsing.Block node) Line 391	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Parsing.Block.Walk(Microsoft.NodejsTools.Parsing.AstVisitor walker) Line 62	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.DDG.Walk(Microsoft.NodejsTools.Parsing.IfNode node) Line 307	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Parsing.IfNode.Walk(Microsoft.NodejsTools.Parsing.AstVisitor visitor) Line 72	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.DDG.Walk(Microsoft.NodejsTools.Parsing.Block node) Line 391	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Parsing.Block.Walk(Microsoft.NodejsTools.Parsing.AstVisitor walker) Line 62	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.DDG.Walk(Microsoft.NodejsTools.Parsing.TryNode node) Line 476	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Parsing.TryNode.Walk(Microsoft.NodejsTools.Parsing.AstVisitor visitor) Line 110	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.DDG.Walk(Microsoft.NodejsTools.Parsing.Block node) Line 391	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Parsing.Block.Walk(Microsoft.NodejsTools.Parsing.AstVisitor walker) Line 62	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.FunctionAnalysisUnit.AnalyzeWorker(Microsoft.NodejsTools.Analysis.Analyzer.DDG ddg, System.Threading.CancellationToken cancel) Line 85	C#
>	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.AnalysisUnit.Analyze(Microsoft.NodejsTools.Analysis.Analyzer.DDG ddg, System.Threading.CancellationToken cancel) Line 177	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.Analyzer.DDG.Analyze(Microsoft.NodejsTools.Analysis.Deque<Microsoft.NodejsTools.Analysis.AnalysisUnit> queue, System.Threading.CancellationToken cancel) Line 72	C#
 	Microsoft.NodejsTools.Analysis.dll!Microsoft.NodejsTools.Analysis.JsAnalyzer.AnalyzeQueuedEntries(System.Threading.CancellationToken cancel) Line 389	C#
 	Microsoft.NodejsTools.dll!Microsoft.NodejsTools.Intellisense.VsProjectAnalyzer.AnalysisQueue.GroupAnalysis.Analyze(System.Threading.CancellationToken cancel) Line 247	C#
 	Microsoft.NodejsTools.dll!Microsoft.NodejsTools.Intellisense.VsProjectAnalyzer.AnalysisQueue.Worker(object threadStarted) Line 183	C#
 	[External Code]	
```

More specifically, analysis stalls somewhere in the middle of all these nested for loops in DefineProperties.
https://github.com/Microsoft/nodejstools/blob/master/Nodejs/Product/Analysis/Analysis/GlobalBuilder.cs#L1311

At this point in time...
```
args[0].Count = 50
args[1].Count = 229
propsObj.Descriptors.Keys.Count = 199
definingProperty.Count = 286
```
Which explains the hang :smiley:. Still need to figure out the fix, but at least this somewhat narrows it down.

**Comment by mousetraps** [comment]
Ahh, we're not checking MaxMergeTypes here... that would explain #505 too

**Comment by eladnava** [comment]
Well done! :+1:

**Comment by mousetraps** [comment]
@eladnava the fix is now available in our [latest dev build](https://github.com/Microsoft/nodejstools/releases). Mind verifying it works fine on your end?

**Comment by eladnava** [comment]
@mousetraps unfortunately, it still hangs :-(

I installed the `NTVS.Dev.2015-10-20.VS.2012.msi` and opened the project with the attached `package.json`, and VS hung again when I wanted to close the solution.

Let me know if there's anything else I can do to help.

**Comment by mousetraps** [comment]
Could you try deleting your ntvs_analysis.dat file first so that analysis will restart on your project?

**Comment by eladnava** [comment]
@mousetraps Since I created the Node.js project from existing files, there is no `.ntvs_analysis.dat` file and code analysis happens on-the-fly when I open the solution, so no file to delete.

It's worth mentioning that only after I edit some files and try to debug does VS hang -- simply opening and closing the solution won't cause this behavior.

**Comment by mousetraps** [comment]
Hmm just to make sure we're on the same page here... you mentioned you were running into the issue on VS 2015. But according to your comment here, it sounds like you installed the VS 2012 msi?

**Comment by eladnava** [comment]
@mousetraps true, I decided to install the VS 2012 NTVS - does it make a difference though? Or was the fix only applied to the VS 2015 dev build?

Also, worth mentioning that I created a brand new project after installing the latest dev build (on VS 2012), so no left over .ntvs_analysis.dat from the previous version.
